ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/tasks/ParallelImportChainSegmentTask.java

```diff
@@ -0,0 +1,192 @@
+              ethContext,
+              ethTasksTimer);
+      final ParallelDownloadBodiesTask<B> downloadBodiesTask =
+          new ParallelDownloadBodiesTask<>(
+              blockHandler,
+              validateHeadersTask.getOutboundQueue(),
+              maxActiveChunks,
+              ethContext,
+              ethTasksTimer);
+      final ParallelValidateAndImportBodiesTask<B> validateAndImportBodiesTask =
+          new ParallelValidateAndImportBodiesTask<>(
+              blockHandler,
+              downloadBodiesTask.getOutboundQueue(),
+              Integer.MAX_VALUE,
+              ethContext,
+              ethTasksTimer);
+
+      // Start the pipeline.
+      final EthScheduler scheduler = ethContext.getScheduler();
+      final CompletableFuture<?> downloadHeaderFuture =
+          scheduler.scheduleServiceTask(downloadHeadersTask);
+      final CompletableFuture<?> validateHeaderFuture =
+          scheduler.scheduleServiceTask(validateHeadersTask);
+      final CompletableFuture<?> downloadBodiesFuture =
+          scheduler.scheduleServiceTask(downloadBodiesTask);
+      final CompletableFuture<AbstractPeerTask.PeerTaskResult<List<List<B>>>> validateBodiesFuture =
+          scheduler.scheduleServiceTask(validateAndImportBodiesTask);
+
+      // Hook in pipeline completion signaling.
+      downloadHeadersTask.setLameDuckMode(true);
+      downloadHeaderFuture.thenRun(() -> validateHeadersTask.setLameDuckMode(true));
+      validateHeaderFuture.thenRun(() -> downloadBodiesTask.setLameDuckMode(true));
+      downloadBodiesFuture.thenRun(() -> validateAndImportBodiesTask.setLameDuckMode(true));
+
+      final BiConsumer<? super Object, ? super Throwable> cancelOnException =
+          (s, e) -> {
+            if (e != null && !(e instanceof CancellationException)) {
+              downloadHeadersTask.cancel();
+              validateHeadersTask.cancel();
+              downloadBodiesTask.cancel();
+              validateAndImportBodiesTask.cancel();
+              result.get().completeExceptionally(e);
+            }
+          };
+
+      downloadHeaderFuture.whenComplete(cancelOnException);
+      validateHeaderFuture.whenComplete(cancelOnException);
+      downloadBodiesFuture.whenComplete(cancelOnException);
+      validateBodiesFuture.whenComplete(
+          (r, e) -> {
+            cancelOnException.accept(r, e);
+            if (r != null) {
+              try {
+                final List<B> importedBlocks =
+                    validateBodiesFuture.get().getResult().stream()
+                        .flatMap(Collection::stream)
+                        .collect(Collectors.toList());
+                result.get().complete(importedBlocks);
+              } catch (final InterruptedException | ExecutionException ex) {
+                result.get().completeExceptionally(ex);
+              }
+            }
+          });
+
+    } else {
+      LOG.warn("Import task requested with no checkpoint headers.");
```

Comment:
We typically switch this around to use an early return (`if (...) { LOG.warn(...); return;}`) rather than nesting the whole method inside the if.

Reply:
this started out life as r and e responses, it got consumed by the cancelOnException lambda.  I'll reform it so it's more clear.